File: user-analytics/billingservice/src/main/java/com/healthcare/billingservice/service/impl/BillingService.java
/*
 * © 2025 Yevhen Tereshchenko
 * All rights reserved.
 *
 */

package com.healthcare.billingservice.service.impl;

import com.healthcare.billingservice.dto.InvoiceDTO;
import com.healthcare.billingservice.dto.PaymentDTO;
import com.healthcare.billingservice.entity.InvoiceStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

public interface BillingService {
    InvoiceDTO createInvoice(InvoiceDTO invoiceDTO);

    @Transactional(readOnly = true)
    InvoiceDTO getInvoiceById(Long id);

    @Transactional(readOnly = true)
    InvoiceDTO getInvoiceByNumber(String invoiceNumber);

    @Transactional(readOnly = true)
    Page<InvoiceDTO> getAllInvoices(Pageable pageable);

    @Transactional(readOnly = true)
    Page<InvoiceDTO> getInvoicesByPatient(Long patientId, Pageable pageable);

    @Transactional(readOnly = true)
    Page<InvoiceDTO> getInvoicesByStatus(InvoiceStatus status, Pageable pageable);

    @Transactional(readOnly = true)
    List<InvoiceDTO> getOverdueInvoices();

    @Transactional(readOnly = true)
    List<InvoiceDTO> getPatientOutstandingInvoices(Long patientId);

    @Transactional(readOnly = true)
    BigDecimal getTotalOutstanding();

    @Transactional(readOnly = true)
    Page<InvoiceDTO> searchInvoices(String searchTerm, Pageable pageable);

    InvoiceDTO updateInvoice(Long id, InvoiceDTO invoiceDTO);

    InvoiceDTO sendInvoice(Long id);

    PaymentDTO addPayment(Long invoiceId, PaymentDTO paymentDTO);

    InvoiceDTO cancelInvoice(Long id);

    void deleteInvoice(Long id);
}

File: user-analytics/healthcare-frontend/src/app/services/billing.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Invoice, Payment, InvoiceStatus, Page } from '../models/billing.model';

@Injectable({
  providedIn: 'root'
})
export class BillingService {
  private readonly API_URL = 'http://localhost:8082/billing-service/api/v1/billing';

  constructor(private http: HttpClient) { }

  private getHeaders(): HttpHeaders {
    const token = localStorage.getItem('token');
    return new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
  }

  createInvoice(invoice: Invoice): Observable<Invoice> {
    return this.http.post<Invoice>(`${this.API_URL}/invoices`, invoice, {
      headers: this.getHeaders()
    });
  }

  getInvoiceById(id: number): Observable<Invoice> {
    return this.http.get<Invoice>(`${this.API_URL}/invoices/${id}`, {
      headers: this.getHeaders()
    });
  }

  getInvoiceByNumber(invoiceNumber: string): Observable<Invoice> {
    return this.http.get<Invoice>(`${this.API_URL}/invoices/number/${invoiceNumber}`, {
      headers: this.getHeaders()
    });
  }

  getAllInvoices(page: number = 0, size: number = 10): Observable<Page<Invoice>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<Invoice>>(`${this.API_URL}/invoices`, {
      headers: this.getHeaders(),
      params
    });
  }

  getInvoicesByPatient(patientId: number, page: number = 0, size: number = 10): Observable<Page<Invoice>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<Invoice>>(`${this.API_URL}/invoices/patient/${patientId}`, {
      headers: this.getHeaders(),
      params
    });
  }

  getInvoicesByStatus(status: InvoiceStatus, page: number = 0, size: number = 10): Observable<Page<Invoice>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<Invoice>>(`${this.API_URL}/invoices/status/${status}`, {
      headers: this.getHeaders(),
      params
    });
  }

  getOverdueInvoices(): Observable<Invoice[]> {
    return this.http.get<Invoice[]>(`${this.API_URL}/invoices/overdue`, {
      headers: this.getHeaders()
    });
  }

  getPatientOutstandingInvoices(patientId: number): Observable<Invoice[]> {
    return this.http.get<Invoice[]>(`${this.API_URL}/invoices/patient/${patientId}/outstanding`, {
      headers: this.getHeaders()
    });
  }

  getTotalOutstanding(): Observable<number> {
    return this.http.get<number>(`${this.API_URL}/invoices/outstanding/total`, {
      headers: this.getHeaders()
    });
  }

  searchInvoices(query: string, page: number = 0, size: number = 10): Observable<Page<Invoice>> {
    const params = new HttpParams()
      .set('query', query)
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<Invoice>>(`${this.API_URL}/invoices/search`, {
      headers: this.getHeaders(),
      params
    });
  }

  updateInvoice(id: number, invoice: Invoice): Observable<Invoice> {
    return this.http.put<Invoice>(`${this.API_URL}/invoices/${id}`, invoice, {
      headers: this.getHeaders()
    });
  }

  sendInvoice(id: number): Observable<Invoice> {
    return this.http.patch<Invoice>(`${this.API_URL}/invoices/${id}/send`, null, {
      headers: this.getHeaders()
    });
  }

  addPayment(invoiceId: number, payment: Payment): Observable<Payment> {
    return this.http.post<Payment>(`${this.API_URL}/invoices/${invoiceId}/payments`, payment, {
      headers: this.getHeaders()
    });
  }

  cancelInvoice(id: number): Observable<Invoice> {
    return this.http.patch<Invoice>(`${this.API_URL}/invoices/${id}/cancel`, null, {
      headers: this.getHeaders()
    });
  }

  deleteInvoice(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/invoices/${id}`, {
      headers: this.getHeaders()
    });
  }
}
